File: app/src/main/java/com/github/abhijitpparate/keeps/screen/home/presenter/HomePresenter.java
package com.github.abhijitpparate.keeps.screen.home.presenter;


import android.util.Log;

import com.facebook.login.LoginManager;
import com.github.abhijitpparate.keeps.data.auth.AuthInjector;
import com.github.abhijitpparate.keeps.data.auth.AuthSource;
import com.github.abhijitpparate.keeps.data.auth.User;
import com.github.abhijitpparate.keeps.data.database.DatabaseInjector;
import com.github.abhijitpparate.keeps.data.database.DatabaseSource;
import com.github.abhijitpparate.keeps.data.database.Note;
import com.github.abhijitpparate.keeps.data.database.Profile;
import com.github.abhijitpparate.keeps.scheduler.SchedulerInjector;
import com.github.abhijitpparate.keeps.scheduler.SchedulerProvider;
import com.google.firebase.auth.FirebaseUser;

import java.util.List;

import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableCompletableObserver;
import io.reactivex.observers.DisposableMaybeObserver;

public class HomePresenter implements HomeContract.Presenter {

    public static final String TAG = "HomePresenter";

    private FirebaseUser currentUser;

    private AuthSource authSource;
    private DatabaseSource databaseSource;
    private SchedulerProvider schedulerProvider;
    private CompositeDisposable disposable;

    private HomeContract.View view;

    boolean isFirstLogin = true;

    public HomePresenter(HomeContract.View view) {
        this.view = view;
        this.authSource = AuthInjector.getAuthSource();
        this.databaseSource = DatabaseInjector.getDatabaseSource();
        this.schedulerProvider = SchedulerInjector.getScheduler();
        this.disposable = new CompositeDisposable();

        view.setPresenter(this);
    }

    @Override
    public void onRefresh() {
        getUserNotesFromDatabase();
    }

    @Override
    public void onLogoutClick() {
        LoginManager.getInstance().logOut();
        disposable.add(
                authSource.logoutUser()
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribeWith(
                                new DisposableCompletableObserver() {
                                    @Override
                                    public void onComplete() {
                                        view.showLoginScreen();
                                        view.makeToast("Logout");
                                    }

                                    @Override
                                    public void onError(@NonNull Throwable e) {
                                        view.makeToast(e.getMessage());
                                    }
                                }
                        )
        );
    }

    @Override
    public void onNewNoteClick(HomeContract.NoteType noteType) {
        view.showNewNoteScreen(noteType);
    }

    @Override
    public void onNoteClick(Note note) {
        view.showNoteScreen(note);
    }

    @Override
    public void subscribe() {
        getUserData();
    }

    private void getUserData() {
//        Log.d(TAG, "getUserData: ");
        view.showProgressBar(true);
        disposable.add(
                authSource
                        .getUser()
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribeWith(new DisposableMaybeObserver<FirebaseUser>() {
                            @Override
                            public void onSuccess(@NonNull FirebaseUser user) {
                                Log.d(TAG, "onSuccess: " + user.getEmail());
                                view.showProgressBar(false);
                                HomePresenter.this.currentUser = user;
                                getUserProfileFromDatabase();
                                getUserNotesFromDatabase();
                            }

                            @Override
                            public void onError(@NonNull Throwable e) {
                                Log.d(TAG, "onError: ");
                                view.showProgressBar(false);
                                view.makeToast(e.getMessage());
                            }

                            @Override
                            public void onComplete() {
                                Log.d(TAG, "onComplete: ");
                                view.showProgressBar(false);
                            }
                        })
        );
    }

    private void getUserNotesFromDatabase() {
        Log.d(TAG, "getUserNotesFromDatabase: ");
        view.showProgressBar(true);
        disposable.add(
                databaseSource
                        .getNotesForCurrentUser(currentUser.getUid())
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribeWith(new DisposableMaybeObserver<List<Note>>() {
                            @Override
                            public void onSuccess(@NonNull List<Note> notes) {
                                view.setNotes(notes);
                                view.showProgressBar(false);
                            }

                            @Override
                            public void onError(@NonNull Throwable e) {
                                view.makeToast(e.getMessage());
                                view.showProgressBar(false);
                            }

                            @Override
                            public void onComplete() {
                                view.showProgressBar(false);
                            }
                        })
        );
    }

    private void getUserProfileFromDatabase(){
        Log.d(TAG, "getUserProfileFromDatabase: ");
        disposable.add(
                databaseSource
                        .getProfile(currentUser.getUid())
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribeWith(new DisposableMaybeObserver<Profile>() {
                            @Override
                            public void onSuccess(@NonNull Profile profile) {
                                Log.d(TAG, "onSuccess: ");
                                if (isFirstLogin){
                                    isFirstLogin = false;
                                    view.setUserInfo(profile);
                                }
                            }

                            @Override
                            public void onError(@NonNull Throwable e) {
                                Log.d(TAG, "onError: ");
                                view.makeToast(e.getMessage());
//                                view.showLoginScreen();
                            }

                            @Override
                            public void onComplete() {
                                Log.d(TAG, "onComplete: ");
//                                view.showLoginScreen();
                            }
                        })
        );
    }

    @Override
    public void unsubscribe() {
        disposable.clear();
    }
}

File: app/src/main/java/com/github/abhijitpparate/keeps/screen/home/presenter/HomePresenter.kt
package com.github.abhijitpparate.keeps.screen.home.presenter


import android.util.Log
import com.facebook.AccessToken
import com.facebook.FacebookSdk
import com.facebook.login.LoginManager
import com.github.abhijitpparate.keeps.data.auth.AuthInjector
import com.github.abhijitpparate.keeps.data.auth.AuthSource
import com.github.abhijitpparate.keeps.data.auth.User
import com.github.abhijitpparate.keeps.data.database.DatabaseInjector
import com.github.abhijitpparate.keeps.data.database.DatabaseSource
import com.github.abhijitpparate.keeps.data.database.Note
import com.github.abhijitpparate.keeps.data.database.Profile
import com.github.abhijitpparate.keeps.scheduler.SchedulerInjector
import com.github.abhijitpparate.keeps.scheduler.SchedulerProvider
import io.reactivex.annotations.NonNull
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableCompletableObserver
import io.reactivex.observers.DisposableMaybeObserver

class HomePresenter(private val view: HomeContract.View) : HomeContract.Presenter {

    lateinit var currentUser: User

    internal var authSource: AuthSource = AuthInjector.authSource
    internal var databaseSource: DatabaseSource = DatabaseInjector.databaseSource
    internal var schedulerProvider: SchedulerProvider = SchedulerInjector.scheduler
    internal var disposable: CompositeDisposable = CompositeDisposable()

    private var isFirstLogin = true

    init {
        view.setPresenter(this)
    }

    override fun onRefresh() {
        getUserNotesFromDatabase()
    }

    override fun onLogoutClick() {
        if (FacebookSdk.isInitialized() && AccessToken.getCurrentAccessToken() != null) {
            LoginManager.getInstance().logOut()
        }
        disposable.add(
                authSource.logoutUser()
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribeWith(
                                object : DisposableCompletableObserver() {
                                    override fun onComplete() {
                                        view.showLoginScreen()
                                        view.makeToast("Logout")
                                    }

                                    override fun onError(@NonNull e: Throwable) {
                                        view.makeToast(e.message.toString())
                                    }
                                }
                        )
        )
    }

    override fun onNewNoteClick(noteType: HomeContract.NoteType) {
        view.showNewNoteScreen(noteType)
    }

    override fun onNoteClick(note: Note) {
        view.showNoteScreen(note)
    }

    override fun subscribe() {
        getUserData()
    }

    private fun getUserData() {
        //        Log.d(TAG, "getUserData: ");
        view.showProgressBar(true)
        disposable.add(
                authSource
                        .retrieveUser()
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribeWith(object : DisposableMaybeObserver<User>() {
                            override fun onSuccess(@NonNull user: User) {
                                Log.d(TAG, "onSuccess: " + user.email)
                                view.showProgressBar(false)
                                this@HomePresenter.currentUser = user
                                getUserProfileFromDatabase()
                                getUserNotesFromDatabase()
                            }

                            override fun onError(@NonNull e: Throwable) {
                                Log.d(TAG, "onError: ")
                                view.showProgressBar(false)
                            }

                            override fun onComplete() {
                                Log.d(TAG, "onComplete: ")
                                view.showProgressBar(false)
                            }
                        })
        )
    }

    private fun getUserNotesFromDatabase() {
        Log.d(TAG, "getUserNotesFromDatabase: ")
        view.showProgressBar(true)
        disposable.add(
                databaseSource
                        .getNotesForCurrentUser(currentUser.uid!!)
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribeWith(object : DisposableMaybeObserver<List<Note>>() {
                            override fun onSuccess(@NonNull notes: List<Note>) {
                                view.setNotes(notes)
                                view.showProgressBar(false)
                            }

                            override fun onError(@NonNull e: Throwable) {
                                view.makeToast(e.message as String)
                                view.showProgressBar(false)
                            }

                            override fun onComplete() {
                                view.showProgressBar(false)
                            }
                        })
        )
    }

    private fun getUserProfileFromDatabase() {
        Log.d(TAG, "getUserProfileFromDatabase: ")
        disposable.add(
                databaseSource
                        .getProfile(currentUser.uid!!)
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribeWith(object : DisposableMaybeObserver<Profile>() {
                            override fun onSuccess(@NonNull profile: Profile) {
                                Log.d(TAG, "onSuccess: ")
                                if (isFirstLogin) {
                                    isFirstLogin = false
                                    view.setUserInfo(profile)
                                }
                            }

                            override fun onError(@NonNull e: Throwable) {
                                Log.d(TAG, "onError: ")
                                view.makeToast(e.message as String)
                                //                                view.showLoginScreen();
                            }

                            override fun onComplete() {
                                Log.d(TAG, "onComplete: ")
                                //                                view.showLoginScreen();
                            }
                        })
        )
    }

    override fun unsubscribe() {
        disposable.clear()
    }

    companion object {
        val TAG = "HomePresenter"
    }
}